drivers/timer/sam0_rtc_timer.c

```diff
@@ -0,0 +1,290 @@
+	RTC0->INTENSET.reg = RTC_MODE0_INTENSET_CMP0;
+#else
+	/* Non-tickless mode uses comparator together with MATCHCLR. */
+	rtc_sync();
+	RTC0->COMP[0].reg = CYCLES_PER_TICK;
+	RTC0->INTENSET.reg = RTC_MODE0_INTENSET_OVF;
+	rtc_counter = 0;
+	rtc_timeout = 0;
+#endif
+
+	/* Enable RTC module. */
+	rtc_sync();
+	RTC0->CTRL.reg |= RTC_MODE0_CTRL_ENABLE;
+
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	ARG_UNUSED(idle);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+	if (ticks < 0 || ticks > MAX_TICKS) {
+		/* Handle K_FOREVER and too big ticks values. */
+		ticks = MAX_TICKS;
+	}
+
+	/* Compute number of RTC cycles until the next timeout. */
+	u32_t count = rtc_count();
+	u32_t timeout = ticks * CYCLES_PER_TICK
+			+ (count - rtc_last) % CYCLES_PER_TICK;
+
+	/* Round to the nearest tick boundary. */
+	timeout = (timeout + CYCLES_PER_TICK - 1) / CYCLES_PER_TICK
+		  * CYCLES_PER_TICK;
+
+	if (timeout < TICK_THRESHOLD) {
+		/* Trigger the interrupt right away if ticks is too low. */
+		NVIC_SetPendingIRQ(DT_RTC_SAM0_IRQ);
+		return;
+	}
+
+	rtc_sync();
+	RTC0->COMP[0].reg = count + timeout;
+
+#else /* !CONFIG_TICKLESS_KERNEL */
+
+	if (ticks < 0) {
+		/* Disable comparator for K_FOREVER and other negative
+		 * values.
+		 */
+		rtc_timeout = rtc_counter;
+		return;
+	}
+
+	if (ticks == 0) {
+		/* Trigger the interrupt right away if ticks is zero. */
+		NVIC_SetPendingIRQ(DT_RTC_SAM0_IRQ);
+		return;
+	}
+
+	/* Avoid race condition between reading counter and ISR incrementing
+	 * it.
+	 */
+	int key = irq_lock();
+
+	rtc_timeout = rtc_counter + ticks;
+	irq_unlock(key);
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+}
+
+/*
+ * Gets number of ticks elapsed since the last z_clock_announce call.
+ */
+u32_t z_clock_elapsed(void)
+{
+#ifdef CONFIG_TICKLESS_KERNEL
+	return (rtc_count() - rtc_last) / CYCLES_PER_TICK;
+#else
+	return rtc_counter - rtc_last;
+#endif
+}
+
+void z_clock_idle_exit(void)
+{
+	/* Nothing to do here, RTC bookkeeping is done in ISR. */
+}
+
+u32_t _timer_cycle_get_32(void)
+{
+	return (u32_t) z_tick_get() * CYCLES_PER_TICK;
+}
+
+/*
+ * Resets RTC to its initial state.
+ */
+void rtc_reset(void)
+{
+	rtc_sync();
+
+	/* Disable interrupt. */
+	RTC0->INTENCLR.reg = RTC_MODE0_INTENCLR_MASK;
+	/* Clear interrupt flag. */
+	RTC0->INTFLAG.reg = RTC_MODE0_INTFLAG_MASK;
+
+	/* Disable RTC module. */
+	RTC0->CTRL.reg &= ~RTC_MODE0_CTRL_ENABLE;
+
+	rtc_sync();
+
+	/* Initiate software reset. */
+	RTC0->CTRL.reg |= RTC_MODE0_CTRL_SWRST;
+}
+
+/*
+ * Handles RTC interrupt.
+ */
+void rtc_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+
+	/* Read and clear the interrupt flag register. */
+	u16_t status = RTC0->INTFLAG.reg;
+
+	RTC0->INTFLAG.reg = status;
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+	/* Just read the current counter and announce the elapsed time. */
+	u32_t count = rtc_count();
+
+	if (count != rtc_last) {
+		z_clock_announce((count - rtc_last) / CYCLES_PER_TICK);
+		rtc_last = count;
```

Comment:
There's a slippage here where cycles between the "official" comparator value and the read from the counter above get added to the delay between two ticks.   Unlikely to be a serious problem in practice, but given that this is a free-running counter you don't need to do that: just increment the announced ticks in rtc_last in units of CYCLES_PER_TICK, it doesn't have to be current time.

Reply:
Definitely a bug, should be fixed now. `rtc_last` changed to hold number of ticks (instead of cycles) in both ticking and tickless modes.